### ISSUE TITLE
Incorrect validation error(s) only using Stax2 writer

### ISSUE THREAD

**Issue by hxc9** [FasterXML/woodstox#16]
Hello,

I've been running into some seriously bizarre behavior with validation in Woodstox. I might be missing something dead-simple, but otherwise I'm inclined to think it's a bug.

The issue is that, taking a given XML document that conforms to a given XSD schema, we get validation exceptions when using _XmlStreamWriter2_ (after calling _validateAgainst_ on the writer), while no exceptions are thrown (as expected) from _XmlStreamReader2_ with the same document and the same schema.

Here is a small example to reproduce, which reads an XML file into _XMLStreamReader2_ then copies it without changes to an _XMLStreamWriter2_ : 

```
import com.ctc.wstx.stax.WstxInputFactory;
import com.ctc.wstx.stax.WstxOutputFactory;
import org.codehaus.stax2.XMLStreamReader2;
import org.codehaus.stax2.XMLStreamWriter2;
import org.codehaus.stax2.validation.XMLValidationSchema;
import org.codehaus.stax2.validation.XMLValidationSchemaFactory;

import javax.xml.stream.XMLStreamException;
import java.io.InputStream;
import java.io.StringWriter;

public class Converter {

    public static void main(String... args) throws XMLStreamException {

        InputStream reader = Converter.class.getClassLoader().getResourceAsStream("test.xml");
        StringWriter writer = new StringWriter();

        XMLValidationSchema schema = XMLValidationSchemaFactory.newInstance(XMLValidationSchema.SCHEMA_ID_W3C_SCHEMA)
                .createSchema(Converter.class.getClassLoader().getResourceAsStream("schema.xsd"));


        XMLStreamReader2 xmlReader = (XMLStreamReader2) new WstxInputFactory().createXMLStreamReader(reader);
        xmlReader.validateAgainst(schema);

        XMLStreamWriter2 xmlWriter = (XMLStreamWriter2) new WstxOutputFactory().createXMLStreamWriter(writer);
        xmlWriter.validateAgainst(schema);

        xmlWriter.copyEventFromReader(xmlReader, false);

        while (xmlReader.hasNext()) {
            xmlReader.next();

            xmlWriter.copyEventFromReader(xmlReader, true);
        }

        System.out.println(writer.toString());
    }
}
```

Here is the XML:

```
<?xml version="1.0" encoding="UTF-8"?>
<JobStatus xsdVersion="NA">
    <Document>
        <DocumentId>1234567890</DocumentId>
    </Document>
    <Document>
        <DocumentId>1234567891</DocumentId>
    </Document>
</JobStatus>
```
and the schema

```
<?xml version="1.0" encoding="UTF-8"?>
<xs:schema elementFormDefault="qualified"
           xmlns:xs="http://www.w3.org/2001/XMLSchema">

    <xs:element name="JobStatus">
        <xs:complexType>
            <xs:sequence>
                <xs:element name="Document" maxOccurs="unbounded">
                    <xs:complexType>
                        <xs:sequence>
                            <xs:element name="DocumentId" type="xs:string"/>
                        </xs:sequence>
                    </xs:complexType>
                </xs:element>
            </xs:sequence>
            <xs:attribute name="xsdVersion" type="xs:string" use="required"/>
        </xs:complexType>
    </xs:element>
</xs:schema>
```
Running the example code with those two files results in the following exception being thrown:

```
Exception in thread "main" com.ctc.wstx.exc.WstxValidationException: element "JobStatus" is missing "xsdVersion" attribute
        at [row,col {unknown-source}]: [1,66]
        at com.ctc.wstx.exc.WstxValidationException.create(WstxValidationException.java:50)
        at com.ctc.wstx.sw.BaseStreamWriter.reportProblem(BaseStreamWriter.java:1223)
        at com.ctc.wstx.msv.GenericMsvValidator.reportError(GenericMsvValidator.java:549)
        at com.ctc.wstx.msv.GenericMsvValidator.reportError(GenericMsvValidator.java:541)
        at com.ctc.wstx.msv.GenericMsvValidator.reportError(GenericMsvValidator.java:535)
        at com.ctc.wstx.msv.GenericMsvValidator.validateElementAndAttributes(GenericMsvValidator.java:343)
        at com.ctc.wstx.sw.BaseNsStreamWriter.closeStartElement(BaseNsStreamWriter.java:420)
        at com.ctc.wstx.sw.BaseStreamWriter.copyEventFromReader(BaseStreamWriter.java:807)
        at Converter.main(Converter.java:34)
```

However, the code executes fine if one removes `xmlWriter.validateAgainst(schema);` and the result XML is identical (modulo white space changes) to the input file, and still conforms to the XSD.

Just for information, I've reproduced this behavior (validation successful on reader, but fails on writer) with different pairs of XML and XSD, in which case you can get different types of error messages at different positions in the document.

So I would say the questions here are:
- why does validation fails, when the document obviously conform to the schema (and when the output of the writer conforms to the schema, if one lets it run by removing the validation) ?
- how can validation of the same document succeed when reading, fails when writing ?

Versions:
tested with Java 7 & 8 using the following dependencies:

- org.codehaus.woodstox stax2-api 4.0.0
- com.fasterxml.woodstox woodstox-core 5.0.2
- net.java.dev.msv msv-core 2013.6.1
- net.java.dev.msv xsdlib 2013.6.1

**Comment by cowtowncoder** [comment]
Most likely this is a bug, caused by a missing call from generator to validation component, and latter not observing an event it should to complete validation. Reader and writer-side (parser, generator) are different pieces of code, so validation failing on one side is not unexpected, just means that one works better than the other; and in this case parser has been much more widely used so problems are more likely to get addressed.

I hope to find time to resolve this, but at the moment there's quite a backlog of things (not for Woodstox, but for Jackson) so this may take a while.

**Comment by cowtowncoder** [comment]
I can reproduce this; added a failing test and hoping to figure this out some time this week, to fix for 5.0.3 if possible.

**Comment by cowtowncoder** [comment]
Turns out the issue was with particular code path related to `copyEventFromReader()`; now fixed.
